### PR TITLE
feat: add aggregation count to SCSI

### PR DESCRIPTION
**Pull request by kaliang1**
This change adds a new method to the DAO and resource to count aggregations. The aspect and path to be grouped/aggregated on can be passed in, and the count of the urns for each grouped value is returned. The returned value is a map of the field value to the count.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-api/src/test/java/com/linkedin/metadata/dao/BaseLocalDAOTest.java [hunk 1]
```diff
@@ -129,6 +130,11 @@ protected void save(FooUrn urn, RecordTemplate value, AuditStamp auditStamp, lon
       return null;
     }
 
+    @Override
+    public Map<String, Long> countAggregate(@Nonnull IndexFilter indexFilter, @Nonnull IndexGroupByCriterion groupCriterion) {
```
Comment (jywadhwani) [inline dao-api/src/test/java/com/linkedin/metadata/dao/BaseLocalDAOTest.java:134]: The original method this is overriding seems to be returning `NonNull` value. Let's return an empty map here instead?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java [hunk 2]
```diff
@@ -1180,4 +1186,103 @@ void addEntityTypeFilter(@Nonnull IndexFilter indexFilter) {
...
+    setCountAggregateParameters(indexCriterionArray, indexGroupByCriterion, query);
+
+    final List<EbeanMetadataIndex> list = query.setDistinct(true).findList();
+    Map<String, Long> result = new HashMap<>();
+    list.stream().map(entry -> {
+      if (entry.getStringVal() != null) {
+        return result.put(entry.getStringVal(), entry.getTotalCount());
+      } else if (entry.getDoubleVal() != null) {
+        return result.put(entry.getDoubleVal().toString(), entry.getTotalCount());
+      } else {
```
Comment (jywadhwani) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:1282]: Let us explicitly check for the `getLongVal` to be non-null. That way if in future we add new columns then this code does not break.

**Review by jywadhwani — CHANGES_REQUESTED** [review]

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+record IndexGroupByCriterion {
+  /**
+   * FQCN of the aspect class in the index table that this criterion refers to e.g. com.linkedin.common.Status
+   */
+  aspect: string
+
+  /**
+   * Corresponding path column of the index table that this criterion refers to e.g. /removed (corresponding to field "removed" of com.linkedin.common.Status aspect)
+   */
+  path: string
```
Comment (jywadhwani) [inline dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl:15]: should this be optional? Do you foresee a case where we may want to group by just the aspect name?

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 4]
```diff
@@ -754,6 +755,16 @@ protected abstract void save(@Nonnull URN urn, @Nonnull RecordTemplate value, @N
...
 
+  /**
+   *  Gets the count of an aggregation specified by the aspect and field to group on.
+   * @param indexFilter {@link IndexFilter} that defines the filter conditions
+   * @param indexGroupByCriterion {@link IndexGroupByCriterion} that defines the aspect to group by
+   * @return map of the field to the count
+   */
+  @Nonnull
+  public abstract Map<String, Long> countAggregate(@Nonnull IndexFilter indexFilter,
+      @Nonnull IndexGroupByCriterion indexGroupByCriterion);
```
Comment (jywadhwani) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:766]: While running SQL query, we can group by multiple columns. This method has provision for grouping only one such column? Do you think we should extend this to multiple columns?

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanLocalDAOTest.java [hunk 5]
```diff
@@ -2078,6 +2071,92 @@ public void testGetWithKeysCount() {
...
+
+    // group by string
+    IndexValue indexValue1 = new IndexValue();
+    indexValue1.setString("val");
+    IndexCriterion criterion1 = new IndexCriterion().setAspect(aspect1)
+        .setPathParams(new IndexPathParams().setPath("/value").setValue(indexValue1).setCondition(Condition.START_WITH));
+
+    IndexCriterionArray indexCriterionArray1 = new IndexCriterionArray(Collections.singletonList(criterion1));
+    final IndexFilter indexFilter1 = new IndexFilter().setCriteria(indexCriterionArray1);
+    IndexGroupByCriterion indexGroupByCriterion1 = new IndexGroupByCriterion().setAspect(AspectFoo.class.getCanonicalName())
```
Comment (jywadhwani) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanLocalDAOTest.java:2110]: let us also test the case when `IndexGroupByCriterion` uses an aspect and path that none of the filtered results contain?

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl (hunk 3 above)
Comment (jywadhwani) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl:15]: Something like: there are 10 filtered results that satisfy `IndexFilter` condition, `aspectA` exists in 8 of them, `aspectB` in 5 of them, etc.

Comment (kaliang1) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl:15]: I think this could potentially be a use case in the future? To group by aspects, perhaps the path can be optional and for the aspects string, we could pass in "*" to indicate that we want to group by all of the aspects. What do you think?

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 4 above)
Comment (kaliang1) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:766]: Sure, I think this could be a use case in the future.

Comment (kaliang1) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:766]: Actually, I'm not sure about how this would look like. For grouping by multiple columns, would that return the count of columns that have the same value for multiple columns? For example, if we group by X, Y, this would put aspects with the same values for both X and Y in one group and return the count?
Also, the returned value may be a nested map, which could get complicated with multiple group by conditions?

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl (hunk 3 above)
Comment (jywadhwani) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl:15]: We could as well have a `Nullable` `IndexGroupByCriterion`? where null value would mean what you are suggesting?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java [hunk 6]
```diff
@@ -980,11 +981,12 @@ private static void validateConditionAndValue(@Nonnull IndexCriterion criterion)
   }
 
   @Nonnull
-  static <ASPECT extends RecordTemplate> String getSortingColumn(@Nonnull IndexSortCriterion indexSortCriterion) {
-    final String[] pathSpecArray = RecordUtils.getPathSpecAsArray(indexSortCriterion.getPath());
+  static <ASPECT extends RecordTemplate> String getFieldColumn(@Nonnull String stringPath,
```
Comment (jywadhwani) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:984]: nit: I would just name these variables as `path` and `aspect` (`string` as prefix doesn't add much value, I think)

Comment (kaliang1) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:984]: For https://github.com/linkedin/datahub-gma/pull/115/files#diff-b5517a7dd054ca33f211c1dddc406b0bebbb4ae5a30cca17c59bff073b221488R989, it uses the variable `aspect` and I think `stringAspect` helps to differentiate them?

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl (hunk 3 above)
Comment (jywadhwani) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/IndexGroupByCriterion.pdl:15]: Let's just skip it for now. We can implement it later. This thread will help with tracking.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 4 above)
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:766]: Sounds good. Let us skip this functionality for now.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java (hunk 6 above)
Comment (jywadhwani) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:984]: how about we use `aspectName` then?

**Review by jywadhwani — APPROVED** [review]
